core/src/main/java/org/apache/gravitino/storage/relational/CachedEntityStore.java

```diff
@@ -0,0 +1,246 @@
+ * under the License.
+ */
+
+package org.apache.gravitino.storage.relational;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Function;
+import org.apache.gravitino.Config;
+import org.apache.gravitino.Entity;
+import org.apache.gravitino.EntityAlreadyExistsException;
+import org.apache.gravitino.EntityStore;
+import org.apache.gravitino.HasIdentifier;
+import org.apache.gravitino.MetadataObject;
+import org.apache.gravitino.NameIdentifier;
+import org.apache.gravitino.Namespace;
+import org.apache.gravitino.SupportsRelationOperations;
+import org.apache.gravitino.cache.EntityCache;
+import org.apache.gravitino.exceptions.NoSuchEntityException;
+import org.apache.gravitino.meta.TagEntity;
+import org.apache.gravitino.tag.SupportsTagOperations;
+import org.apache.gravitino.utils.Executable;
+
+/** * Cached Entity store, which caches metadata in memory. */
+public class CachedEntityStore
+    implements EntityStore, SupportsTagOperations, SupportsRelationOperations {
+  private final RelationalEntityStore entityStore;
+
+  private final EntityCache cache;
+
+  /**
+   * Constructs a new instance of CachedEntityStore.
+   *
+   * @param entityStore The {@link RelationalEntityStore} instance to do the real work.
+   * @param cache The {@link EntityCache} instance to cache metadata in memory.
+   */
+  public CachedEntityStore(EntityStore entityStore, EntityCache cache) {
+    Preconditions.checkArgument(
+        entityStore instanceof RelationalEntityStore,
+        "CachedEntityStore requires a RelationalEntityStore instance.");
+
+    this.entityStore = (RelationalEntityStore) entityStore;
+    this.cache = cache;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void initialize(Config config) throws RuntimeException {
+    entityStore.initialize(config);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public <E extends Entity & HasIdentifier> List<E> list(
+      Namespace namespace, Class<E> type, Entity.EntityType entityType) throws IOException {
+    return cache.withCacheLock(
+        () -> {
+          List<E> entities = entityStore.list(namespace, type, entityType);
+          entities.forEach(cache::put);
+
+          return entities;
+        });
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public <E extends Entity & HasIdentifier> List<E> list(
+      Namespace namespace, Class<E> type, Entity.EntityType entityType, boolean allFields)
+      throws IOException {
+    return cache.withCacheLock(
+        () -> {
+          List<E> entities = entityStore.list(namespace, type, entityType, allFields);
+          entities.forEach(cache::put);
+
+          return entities;
+        });
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean exists(NameIdentifier ident, Entity.EntityType entityType) throws IOException {
+    return cache.withCacheLock(
+        () -> {
+          if (cache.contains(ident, entityType)) {
```

Comment:
I saw you already have a same lock in cache, and you're using the lock again in here. So is it necessary to have a lock in the cache class, since you will always have the lock here?